### ISSUE TITLE
Parallelize Alternator tests

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#14387]
Today, test.py runs Alternator tests using `test/alternator/run`. This runs Scylla once in the background, and then runs pytest asking it to run more than 650 test functions against this running Scylla. These test functions are run sequentially, one after another.

Clearly, running the tests sequentially with no parallelization at all isn't ideal - some of the tests have (small) delays, during which the Scylla process does nothing, so running several tests in parallel could have used the Scylla process to the fullest and finished the entire test suite faster. Until recently nobody perceived this as a real problem - Alternator tests were written to be very fast - today, with a "dev"-mode build on a typical laptop, the entire suite of 650 test functions takes only 25 seconds. No developer has ever complained that test/alternator/run is too slow for development work (and moreover, it also allows running only a single test file or even a single test function, so even that 25 seconds is rarely experienced).

However, as @alecco noted in #13905, the test/alternator/run speed becomes more of a problem in debug build mode, which can be 30 times (!) slower than dev build mode, and take around 13 minutes. This can be frustrating for a developer to see this test suite take 13 minutes to complete, all the time using just one CPU (for handling requests coming from a single test thread) while this user maybe has many CPUs. Individual long tests can also be problematic for test.py for other reasons (that can be argued to be test.py bugs, but still).

An easy and effective way to fix this problem is:

* Have test.py run multiple Scylla processes, and run each `test/alternator/*.py` test file against a different process.

@kostja already did exactly this for cql-pytest: test.py used to run test/cql-pytest/run but today it runs multiple Scylla processes and invokes individual pytest on individual test files against those Scylla processes. Doing this also for Alternator tests should be fairly easy, perhaps even trivial.

This parallelization approach is ideal for CPU-dominated runs (like the debug build) because each of the CPUs can be fully utilized running the slow code on a Scylla of its own. 

However, this parallelization approch is less ideal for a delay-dominated run, where each of the Scylla processes will still have a lot of delays with zero CPU usage, so we will need more Scylla processes than CPUs, and waste a lot of memory. If we want to solve this problem too (I think we probably don't!) we can consider two other parallelization approaches:

* Run a single Scylla process on N cpus, and run multiple pytest invocations in parallel (e.g., run individual test files with separate pytest invocations) against the same Scylla. This should be easy to implement in the test/alternator/run script, and I expect it to mostly work perhaps except a few tests which modify the global Scylla state (e.g., enable tracing) - those were usually written to tolerate other things happening in parallel, but maybe we missed something.
* Run a single Scylla process, but run a single pytest invocation that runs test cases in parallel. This will be harder to implement, and we can run into more test bugs (most tests should tolerate concurrency, but I guess some have bugs).

As noted above, I propose in this issue that we take the first approach, not the latter two.

**Comment by nyh** [comment]
As @mykaul suggested in #13905, let's do both the bullet point 1 (many-Scylla - many-pytest, test.py-style solution) and bullet point 2 (single-Scylla many-pytest solution, inside test/alternator/run). We won't do bullet point 3 (single-Scylla single-pytest).

**Comment by nyh** [comment]
I implemented the second bullet-point solution (single Scylla, many pytest invocations in parallel), and starting testing it on dev-mode build (which, admittedly, is not the case that started this discussion in #13905). The results are reasonable, but not great:

1. Just splitting the test run to pytest separately on each test file instead of one pytest run, increased the total (serial) runtime from 25 seconds to 44 seconds. The extra waste is pytest startup and teardown time and extra work on creating various fixtures shared between multiple test files (like the test tables).
2. Running those pytest invocations in parallel instead of serially brings the total test time down to 17 seconds. This is nice, but not amazing speedup compared to the number we started with (25). As usual in such attempts, we run afoul of Amdahl's Law - one of the last test files is longer than the rest and finishes last alone.
3. Some tests become flaky, we'll need to fix that. In particular it appears that because test table names are picked based on the current time, running multiple tests in parallel can choose the same test name. Maybe we should add the current pid to the test table name.
4. In once case, I got Scylla to crash, on` wasmtime_runtime::traphandlers::unix::trap_handler at wasmtime_runtime.550a4292-cgu.11:?`. This is probably a real bug that should be fixed. This only happened once.

**Comment by nyh** [comment]
https://github.com/scylladb/scylladb/issues/18188 is about doing the first bullet-point, which is also the option I suggested to do above. So if that issue is done, this one can be closed too.

**Comment by xtrey** [comment]
![image (3)](https://github.com/scylladb/scylladb/assets/2657095/259c14a0-d7fa-44e9-9cd0-7e8213836bdd)
Executed the alternator test migrated to PythonSuite with 4 threads. Seems the further increase of the threads will not help, since we have one class that lasts too long and the whole run is limited by it.

**Comment by xtrey** [comment]
The previous result should be ignored. One of the Scylla parameters where incorrect and it was affecting performance so drastically.
Here are the correct results. 
![image](https://github.com/scylladb/scylladb/assets/2657095/2d0a6b6a-48d4-416c-8726-0e4d068fd6a9)

**Comment by nyh** [comment]
I'm closing this issue, as fixed by @xtrey in scylladb/scylladb#18442.

That pull request uses the approach in the first bullet point in the first comment, i.e., the same thing that was done for cql-pytest: test.py runs each of the different files in test/alternator separately. It is perhaps not the most optimal way of parallelizing, but this is the best that test.py has to offer right now, and is good enough for now. And if we later improve it, the mechanism probably won't be specific to Alternator so we don't need a separate issue for the Alternator case..
